### PR TITLE
Replace dask 'compute()' usage with a common realisation call. (#2)

### PR DESCRIPTION
**Pull request by lbdreyer**
Replaces #2422

### CODE REVIEW

File: lib/iris/tests/unit/analysis/test_VARIANCE.py [hunk 1]
```diff
@@ -23,13 +23,13 @@
...
 import iris.tests as tests
 
-from iris._lazy_data import as_lazy_data
 import numpy as np
 import numpy.ma as ma
 
 from iris.analysis import VARIANCE
 import iris.cube
 from iris.coords import DimCoord
+from iris._lazy_data import as_lazy_data, as_concrete_data
```
Comment (lbdreyer) [inline lib/iris/tests/unit/analysis/test_VARIANCE.py]: We do need to agree on an order for importing hidden modules, atm we're inconsistent with it. 
Unfortunately, a quick google didn't unearth any standard practice
I do like this approach of putting them in alphabetical, regardless of whether or not they are private though

File: lib/iris/_lazy_data.py [hunk 2]
```diff
@@ -73,6 +73,30 @@ def as_lazy_data(data, chunks=_MAX_CHUNK_SIZE):
...
+def as_concrete_data(data, nans_replacement=None, result_dtype=None):
+    """
+    Return the actual content of a lazy array, as a numpy array.
+
+    If the data is a NumPy array, return it unchanged.
+
+    If the data is lazy, return the realised result.
+
+    Where lazy data contains NaNs these are translated by filling or conversion
+    to masked data, using the :func:`convert_nans_array` function.
```
Comment (lbdreyer) [inline lib/iris/_lazy_data.py]: >:func:`convert_nans_array`
should this be `:func:`iris._lazy_data.convert_nans_array or is okay to be relative if it's in the same module?

File: lib/iris/_lazy_data.py [hunk 3]
```diff
@@ -73,6 +73,30 @@ def as_lazy_data(data, chunks=_MAX_CHUNK_SIZE):
     return data
 
 
+def as_concrete_data(data, nans_replacement=None, result_dtype=None):
+    """
+    Return the actual content of a lazy array, as a numpy array.
+
+    If the data is a NumPy array, return it unchanged.
```
Comment (bjlittle) [inline lib/iris/_lazy_data.py]: @lbdreyer This also applies to `~numpy.ma.core.MaskedArray`.

File: lib/iris/_merge.py [hunk 4]
```diff
@@ -33,7 +33,8 @@
 import numpy as np
 import numpy.ma as ma
 
-from iris._lazy_data import as_lazy_data, is_lazy_data, multidim_lazy_stack
+from iris._lazy_data import (as_lazy_data, is_lazy_data, multidim_lazy_stack,
+                             as_concrete_data)
```
Comment (bjlittle) [inline lib/iris/_merge.py]: @lbdreyer The sort order needs changes here ... make `as_concrete_data` first ...

File: lib/iris/tests/unit/analysis/test_VARIANCE.py (hunk 1 above)
Comment (bjlittle) [inline lib/iris/tests/unit/analysis/test_VARIANCE.py]: If we're going to go with alphabetical (which we are) then `_` sorts before all lower case letters, so IMHO private modules should go first.

File: lib/iris/_lazy_data.py (hunk 2 above)
Comment (bjlittle) [inline lib/iris/_lazy_data.py]: You could always make it 
```python
:func:`~iris._lazy_data.convert_nans_array`
```

File: lib/iris/coords.py [hunk 5]
```diff
@@ -39,6 +39,7 @@
 
 import iris.aux_factory
 import iris.exceptions
+from iris._lazy_data import as_concrete_data
```
Comment (bjlittle) [inline lib/iris/coords.py]: Import order ...

File: lib/iris/coords.py [hunk 6]
```diff
@@ -1611,7 +1612,11 @@ def _sanitise_array(self, src, ndmin):
     def points(self):
         """Property containing the points values as a numpy array"""
         if is_lazy_data(self._points):
-            self._points = self._points.compute()
+            self._points = as_concrete_data(self._points,
+                                            nans_replacement=np.ma.masked)
+            # NOTE: we probably don't have full support for masked aux-coords.
+            # We certainly *don't* handle a _FillValue attribute (and possibly
+            # the loader will throw one away ?)
```
Comment (bjlittle) [inline lib/iris/coords.py]: @lbdreyer We should raise a ticket to consider how we deal with masked integral data on coordinates and cell measures. At the moment we don't keep the result dtype ... this is lost in translation, which is bad.

File: lib/iris/cube.py [hunk 7]
```diff
@@ -43,7 +43,8 @@
 import iris._constraints
 from iris._deprecation import warn_deprecated
 from iris._lazy_data import (array_masked_to_nans, as_lazy_data,
-                             convert_nans_array, is_lazy_data)
+                             convert_nans_array, is_lazy_data,
+                             as_concrete_data)
```
Comment (bjlittle) [inline lib/iris/cube.py]: Import sort order ...

File: lib/iris/fileformats/pp.py [hunk 8]
```diff
@@ -43,7 +43,8 @@
 import iris.fileformats.rules
 import iris.fileformats.pp_rules
 import iris.coord_systems
-from iris._lazy_data import array_masked_to_nans, as_lazy_data, is_lazy_data
+from iris._lazy_data import (array_masked_to_nans, as_lazy_data, is_lazy_data,
+                             as_concrete_data)
```
Comment (bjlittle) [inline lib/iris/fileformats/pp.py]: Import sort order ...

File: lib/iris/tests/unit/analysis/test_STD_DEV.py [hunk 9]
```diff
@@ -23,7 +23,7 @@
 # importing anything else.
 import iris.tests as tests
 
-from iris._lazy_data import as_lazy_data
+from iris._lazy_data import as_lazy_data, as_concrete_data
```
Comment (bjlittle) [inline lib/iris/tests/unit/analysis/test_STD_DEV.py]: Import sort order

File: lib/iris/tests/unit/lazy_data/test_multidim_lazy_stack.py [hunk 10]
```diff
@@ -26,7 +26,7 @@
 import numpy as np
 import dask.array as da
 
-from iris._lazy_data import as_lazy_data, multidim_lazy_stack
+from iris._lazy_data import as_lazy_data, multidim_lazy_stack, as_concrete_data
```
Comment (bjlittle) [inline lib/iris/tests/unit/lazy_data/test_multidim_lazy_stack.py]: Import sort order

**Comment by lbdreyer** [comment]
@bjlittle Hopefully you are happy with the changes? 
(I am happy for you to squash and merge)

File: lib/iris/_lazy_data.py [hunk 11]
```diff
@@ -73,6 +73,31 @@ def as_lazy_data(data, chunks=_MAX_CHUNK_SIZE):
...
+    Return the actual content of a lazy array, as a numpy array.
+
+    If the data is a NumPy ndarray or masked array, return it unchanged.
+
+    If the data is lazy, return the realised result.
+
+    Where lazy data contains NaNs these are translated by filling or conversion
+    to masked data, using the :func:`~iris._lazy_data.convert_nans_array`
+    function.
+    See there for usage of the 'nans_replacement' and 'result_dtype' keys.
```
Comment (dkillick) [inline lib/iris/_lazy_data.py]: Typo: "there" --> "their".

Alternatively you could change the signature to 

```python
def as_concrete_data(data, **kwargs)
```

as the `kwargs` are not used at all by this function and are just passed straight to `convert_nans_array`. At that point you could replace this description of the `kwargs` with something like

"Kwargs are passed straight to :func:`~iris._lazy_data.convert_nans_array`."

It is a more typical way of doing such things in Python but also less readable...

File: lib/iris/_merge.py [hunk 12]
```diff
@@ -1217,10 +1218,11 @@ def merge(self, unique=True):
...
                 # All inputs were concrete, so turn the result back into a
                 # normal array.
-                merged_data = merged_data.compute()
-                # Unmask the array only if it is filled.
+                merged_data = as_concrete_data(merged_data,
+                                               nans_replacement=ma.masked)
+                # Unmask the array if it has no masked points.
                 if (ma.isMaskedArray(merged_data) and
-                        ma.count_masked(merged_data) == 0):
+                        not ma.is_masked(merged_data)):
```
Comment (dkillick) [inline lib/iris/_merge.py]: Nice spot :wink:

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        lazy_array = as_lazy_data(data)
+        sentinel = mock.sentinel.data
+        with mock.patch('iris._lazy_data.convert_nans_array') as conv_nans:
+            conv_nans.return_value = sentinel
+            result = as_concrete_data(lazy_array, nans_replacement=7,
+                                      result_dtype=np.int16)
+        self.assertEqual(sentinel, result)
+
+        # Check call to convert_nans_array
+        conv_nans.assert_called_once()
```
Comment (dkillick) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: I reckon you can do [`assert_called_once_with`](http://www.voidspace.org.uk/python/mock/mock.html#mock.Mock.assert_called_once_with), which may be useful here.

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            result = as_concrete_data(lazy_array)
+        self.assertEqual(sentinel, result)
+
+        # Check call to convert_nans_array
+        conv_nans.assert_called_once()
+        args, kwargs = conv_nans.call_args
+        arg, = args
+        self.assertFalse(is_lazy_data(arg))
+        self.assertArrayEqual(arg, data)
+        self.assertEqual(kwargs, {'result_dtype': None,
```
Comment (dkillick) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: `assert_called_once_with`?

File: lib/iris/coords.py (hunk 6 above)
Comment (bjlittle) [inline lib/iris/coords.py]: @lbdreyer Did you create an issue to cover this?

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+# Import iris.tests first so that some things can be initialised before
+# importing anything else.
+import iris.tests as tests
+
+import numpy as np
+import numpy.ma as ma
+import dask.array as da
```
Comment (bjlittle) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: @lbdreyer Sorry, but could you fix this import order ...

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py (hunk 14 above)
Comment (lbdreyer) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: Doesn't work when one of the args is a numpy array

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py (hunk 15 above)
Comment (lbdreyer) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: Ah man! I didn't see this one!

File: lib/iris/_lazy_data.py (hunk 11 above)
Comment (lbdreyer) [inline lib/iris/_lazy_data.py]: I think "See there for usage" was meant as in "See over there for usage". 

I like your idea of `**kwargs`

File: lib/iris/coords.py (hunk 6 above)
Comment (lbdreyer) [inline lib/iris/coords.py]: No, but I'll do that now

Comment (lbdreyer) [inline lib/iris/coords.py]: @bjlittle I've just raised https://github.com/SciTools/iris/issues/2449
 not sure if that's quite what you were after. Feel free to edit the issue

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py (hunk 14 above)
Comment (dkillick) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: It probably would if you also mocked the array.

File: lib/iris/_lazy_data.py (hunk 11 above)
Comment (dkillick) [inline lib/iris/_lazy_data.py]: > I think "See there for usage" was meant as in "See over there for usage".

Ah yes, on a re-reading you're quite right!

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py (hunk 14 above)
Comment (dkillick) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: Also, in this case, neither of them are NumPy arrays...

File: lib/iris/tests/unit/lazy_data/test_as_concrete_data.py (hunk 13 above)
Comment (bjlittle) [inline lib/iris/tests/unit/lazy_data/test_as_concrete_data.py]: @dkillick I don't think that `assert_called_once_with` will cope with comparing numpy arrays, so checking the call count and unpacking the call args seems okay ... but you could give @dkillick's suggestion a try.

@lbdreyer Although, you could combine this test and the above `test_lazy_data` test into one and also use sentinels for the `nans_replacement` and `result_dtype` ... that would make it sufficiently generic.

**Comment by bjlittle** [comment]
:+1: Hoo-rah! Awesome, thanks @lbdreyer

**Comment by lbdreyer** [comment]
Thanks @bjlittle !!

**Pull request merged**